### DISCUSSION TITLE
When should a function/address should be payable?

### DISCUSSION THREAD

**Discussion by NotNotRama** [Cyfrin/foundry-full-course-cu#511] [Q&A] [answered]
In early lessons this was covered when we were using `call` (for example, `(bool success,) = winner.call{value: address(this).balance}("")`) but in lesson 12, the DSC engine also manages transfers but functions don't require the `payable` keyword. Like here:

```js
    function depositCollateral(address tokenCollateralAddress, uint256 amountCollateral)
        public
        moreThanZero(amountCollateral)
        isAllowedToken(tokenCollateralAddress)
        nonReentrant
    {
        s_collateralDeposited[msg.sender][tokenCollateralAddress] += amountCollateral;
        emit CollateralDeposited(msg.sender, tokenCollateralAddress, amountCollateral);
        bool success = IERC20(tokenCollateralAddress).transferFrom(msg.sender, address(this), amountCollateral);

        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }
  ```

So when is that an address or function should be marked as `payable`?

**Comment by shikhar229169** [comment]
Payable keyword is required when we are sending eth.
But over here we are actually not sending eth, we are sending the wETH and this is a token which has its own contract and also its own methods, you can check the wETH contract at https://sepolia.etherscan.io/address/0xdd13E55209Fd76AfE204dBda4007C227904f0a81

You can understand it like:
- To deposit the collateral on DSCEngine you need eth in form of wrapped ETH which is actually a token.
- And to get  wETH, you need to send actual eth to the wETH contract and in returns it gives you wETH (you can get your actual eth back by giving the wETH back to the wETH contract).
- Now you have eth in form of tokens (that is not real eth), it is just token which you got by depositing eth on wETH contract.
- Before the depositCollateral function can be called on DSCEngine contract you are first required to approve the DSCEngine with the required wETH tokens (which you need to deposit on DSCEngine) via the wETH contract.
- While transferring the wETH, actually there is no transfer of real eth being taken place because it is just a token transfer we are not required to mark the addresses as payable.

**Reply by NotNotRama** [reply]
Think I get it but now I'm wondering, why is that we're using weth instead of just eth in the contract? Also, what's the purpose of weth? Is it because we might not be on the mainnet and we need a way to operate with eth?

**Reply by shikhar229169** [reply, marked as answer]
Nah! Its not related to mainnet in any way, here we are using sepolia which is not related to mainnet.
For sure we can have it in form of eth.
But, in the course we are using wETH.
wETH is a token as I discussed. Its like you deposit real ETH on the wETH contract and you get equivalent amount of wrapped eth from there. Note that both are different thing, but they have same value.

There are various protocols which don't support eth for exchanges, they support only weth so this is one of the significance. (like our Decentralized Stable Coin protocol).
